students/douglas_klos/session9/mailroom/donordb.py

```diff
@@ -0,0 +1,247 @@
+            return_string += f'{self.database[key].name:>24} : {self.database[key].donations}\n'
+        return return_string
+
+    def add_donor(self, donor):
+        """
+        Add a donor to the database
+
+        :param donor: Donor to be added to database. Donor objects will be added directly,
+                          names will be added as donor objects with no donations.
+        """
+        if isinstance(donor, Donor):
+            self.database[donor.name] = donor
+        elif donor in self.database.keys():
+            return f'{donor} already exists in database'
+        else:
+            self.database[donor] = Donor(donor)
+
+        return f'{donor} has been added to the database'
+
+    def add_donation(self, donor, donation):
+        """
+        Add a donation to the database
+
+        :param donor: Name of donor to add donation to
+        :param donation: Donation to be added to donor
+        """
+        try:
+            float(donation)
+        except ValueError:
+            return f'{donation} is not a valid donation amount'
+
+        if donation <= 0:
+            return f'{donation} is not a valid donation amount'
+
+        try:
+            self.database[donor].add_donation(donation)
+            return f'Donation {donation} has been added to donor {donor}'
+        except KeyError:
+            return f'{donor} not found in database'
+
+    def remove_donor(self, donor):
+        """
+        Remove a donor from the database
+
+        :param donor: Donor to be removed from database
+        """
+        try:
+            del self.database[donor]
+            return f'{donor} removed from database'
+        except KeyError:
+            return f'{donor} not found in database'
+
+    def remove_donation(self, donor, donation):
+        """
+        Remove a donation from the database
+
+        :param donor: Name of donor to remove donation from
+        :param donation: Donation to be added to the database
+        """
+        try:
+            return self.database[donor].remove_donation(donation)
+        except KeyError:
+            return f'Donation {donation} from donor {donor} not found in database'
+
+    def display_report(self):
+        """ Prints a report of donors and their donations """
+        report = '\n' + "-" * 79 + '\n'
+        report += 'Donor Name\t\t|           Total Given | Num Gifts |      Average Gift\n'
+        report += "-" * 79 + '\n'
+
+        for key in self.database:
+            report += (f'{self.database[key].name:24s}   '
+                       f'$ {self.database[key].total_donations:18,.2f}')
+            if self.database[key].donations == []:
+                report += f'\t\t0    '
+            else:
+                report += f'{len(self.database[key].donations):10}    '
+            report += f'$ {self.database[key].average_donation:14,.2f}\n'
+
+        report += "-" * 79 + '\n'
+
+        return report
+
+    def html_report(self):
```

Comment:
Fancy :-) 
This is good, still, any data visualization code should be separated from the data itself.